File: pkg/controller/client/type.go
package client

import (
	"context"
	"net/http"

	"inspr.dev/inspr/pkg/api/models"
	"inspr.dev/inspr/pkg/meta"
	"inspr.dev/inspr/pkg/meta/utils/diff"
	"inspr.dev/inspr/pkg/rest"
	"inspr.dev/inspr/pkg/rest/request"
)

// TypeClient is a client for manipulating Type structures in Insprd
type TypeClient struct {
	reqClient *request.Client
}

// Get gets a Type from Insprd, if it exists.
// The scope refers to the dApp in which the Type is in, represented with a dot separated query
// such as app1.app2
func (tc *TypeClient) Get(ctx context.Context, scope, name string) (*meta.Type, error) {
	tdi := models.TypeQueryDI{
		TypeName: name,
	}
	var resp meta.Type

	err := tc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/types", http.MethodGet, request.DefaultHost, tdi, &resp)
	if err != nil {
		return nil, err
	}

	return &resp, nil
}

// Create creates given Type inside of Insprd
// The scope refers to the dApp in which the Type will be in, represented with a dot separated query
// such as app1.app2
func (tc *TypeClient) Create(ctx context.Context, scope string, t *meta.Type, dryRun bool) (diff.Changelog, error) {
	tdi := models.TypeDI{
		Type:   *t,
		DryRun: dryRun,
	}
	var resp diff.Changelog

	err := tc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/types", http.MethodPost, request.DefaultHost, tdi, &resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

// Delete deletes a Type from Insprd, if exists.
// The scope refers  to the dApp in which the Type is in, represented with a dot separated query
// such as app1.app2. The name is the name of the Type to be deleted
func (tc *TypeClient) Delete(ctx context.Context, scope, name string, dryRun bool) (diff.Changelog, error) {
	tdi := models.TypeQueryDI{
		TypeName: name,
		DryRun:   dryRun,
	}
	var resp diff.Changelog

	err := tc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/types", http.MethodDelete, request.DefaultHost, tdi, &resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

// Update updates given Type in Insprd, if it exists.
// The scope refers to the dApp in which the Type is in, represented with a dot separated query
// such asapp1.app2
func (tc *TypeClient) Update(ctx context.Context, scope string, t *meta.Type, dryRun bool) (diff.Changelog, error) {
	tdi := models.TypeDI{
		Type:   *t,
		DryRun: dryRun,
	}
	var resp diff.Changelog

	err := tc.reqClient.
		Header(rest.HeaderScopeKey, scope).
		Send(ctx, "/types", http.MethodPut, request.DefaultHost, tdi, &resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

File: clients/python/inspr/controller/type.py
import sys
from ..models import *
from ..rest import *
from .client import *

TYPE_ROUTE = "types"

class TypeClient(Client):
    def get(self, scope:str, type_name:str) -> InsprStructure:
        msg_body = {
            "typename": type_name
        }
        
        headers = self.get_header_with_scope(scope)

        try:
            resp = send_get_request(self.url + "/" + TYPE_ROUTE, body=msg_body, headers=headers)
            return InsprStructure(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Get Type request: {e}")

    def delele(self, scope:str, type_name:str, dryRun:bool) -> Changelog:
        msg_body = {
            "typename": type_name,
            "dry": dryRun
        }

        headers = self.get_header_with_scope(scope)

        try:
            resp = send_delete_request(self.url + "/" + TYPE_ROUTE, body=msg_body, headers=headers)
            return Changelog(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Delete Type request: {e}")

    def create(self, scope:str, type:dict, dryRun:bool) -> Changelog:
        msg_body = {
            "type": type,
            "dry": dryRun
        }

        headers = self.get_header_with_scope(scope)

        try:
            resp = send_post_request(self.url + "/" + TYPE_ROUTE, msg_body, headers)
            return Changelog(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Create Type request: {e}")

    def update(self, scope:str, type:dict, dryRun:bool) -> Changelog:
        msg_body = {
            "type": type,
            "dry": dryRun
        }

        headers = self.get_header_with_scope(scope)

        try:
            resp = send_update_request(self.url + "/" + TYPE_ROUTE, msg_body, headers)
            return Changelog(json.loads(resp.text))
        except Exception as e:
            raise Exception(f"Error while send a Update Type request: {e}")